pkg/controller/daemon/update.go

```diff
@@ -194,6 +220,24 @@ func (dsc *DaemonSetsController) rollingUpdate(ctx context.Context, ds *apps.Dae
 	// use any of the candidates we can, including the allowedNewNodes
 	logger.V(5).Info("DaemonSet allowing replacements", "daemonset", klog.KObj(ds), "replacements", len(allowedNewNodes), "maxSurge", maxSurge, "numSurge", numSurge, "candidates", len(candidateNewNodes))
 	remainingSurge := maxSurge - numSurge
+
+	if remainingSurge > 0 {
+
+		diffAvailableAndDesired := int(getNumAvailablePod(ds, nodeToDaemonPods, now) - ds.Status.DesiredNumberScheduled)
+
+		if diffAvailableAndDesired < 0 {
+			diffAvailableAndDesired = 0
+		}
+		if numPodToDeletePod := len(shouldNotRunPodsToDelete); diffAvailableAndDesired > numPodToDeletePod {
+			diffAvailableAndDesired = numPodToDeletePod
+		}
+		if diffAvailableAndDesired > remainingSurge {
```

Comment:
why do we need to consult remaining surge? We can delete these pods even when remainingSurge is 0. Because these pods are living on the extra non schedulable nodes...

Reply:
Oops, sorry!
I forgot to delete the implementation I wrote as a test...
I'll remove that.